File: android/app/src/main/java/org/switchboard/android/irc/Editing.kt
package org.switchboard.android.irc

/**
 * Which of your own messages you can still change.
 *
 * The Kotlin half of `src/shared/editing.ts`. Both clients offered Edit in two
 * places it does not work: on a server that does not carry edits, where the
 * new text arrives as a *second message* because an edit is only a `PRIVMSG`
 * wearing a client tag; and on an action, where `/me waves` is CTCP-wrapped
 * and the edit sends plain text, turning something you did into something you
 * said.
 */
object Editing {

    /**
     * Both spellings. A capability loses its `draft/` prefix when it is
     * ratified, and a client that only knows the draft name stops offering the
     * feature on the day the servers stop being drafts.
     */
    private val EDIT_CAPS = setOf("draft/message-edit", "message-edit")

    fun editsAllowed(capabilities: Collection<String>): Boolean =
        capabilities.any { it in EDIT_CAPS }

    /** Whether this message is yours, and of a kind an edit would not spoil */
    fun canEdit(
        type: String,
        nick: String,
        deleted: Boolean,
        myNick: String,
        capabilities: Collection<String>
    ): Boolean {
        if (!editsAllowed(capabilities)) return false
        // Nothing to amend, and the server would refuse the attempt
        if (deleted) return false
        // Plain text only: an action, a notice and a system line each mean
        // something the edit would not carry.
        if (type != "privmsg") return false
        if (myNick.isEmpty()) return false
        return nick.lowercase() == myNick.lowercase()
    }
}

File: src/shared/editing.ts
/**
 * Which of your own messages you can still change.
 *
 * Both clients offer an Edit action and neither asked this properly, so both
 * offered it in two places it does not work:
 *
 *  - **On a server that does not carry edits.** An edit is an ordinary
 *    `PRIVMSG` wearing a `+draft/edit` client tag. A server without the
 *    capability relays it as what it looks like — so "editing" a line on one
 *    of those posts the new text as a *second message* and leaves the first
 *    where it was. Worse than refusing.
 *  - **On an action.** `/me waves` is a `PRIVMSG` with the text wrapped in
 *    CTCP, and the edit sends the box's contents as plain text — so amending
 *    one turned it from something you did into something you said.
 *
 * Shared because the question is the same on both, and because the desktop
 * asks it a third time: the Up key opens the last message you can edit.
 */

/** As much of a message as this question needs */
export interface EditableMessage {
  readonly id: string
  readonly nick: string
  readonly type: string
  readonly deleted?: boolean
}

/**
 * Both spellings.
 *
 * Capabilities lose their `draft/` prefix when they are ratified, and a client
 * that only knows the draft name stops offering the feature on the day the
 * servers stop being drafts — see `isupportValue`, which learned the same
 * lesson about tokens.
 */
const EDIT_CAPS = ['draft/message-edit', 'message-edit']

export function editsAllowed(capabilities: readonly string[]): boolean {
  return capabilities.some((capability) => EDIT_CAPS.includes(capability))
}

/** Whether this message is yours, and of a kind an edit would not spoil */
export function canEdit(
  message: EditableMessage,
  myNick: string,
  capabilities: readonly string[]
): boolean {
  if (!editsAllowed(capabilities)) return false
  // Nothing to amend, and the server would refuse the attempt
  if (message.deleted) return false
  // Plain text only: an action, a notice and a system line each mean something
  // the edit would not carry.
  if (message.type !== 'privmsg') return false
  if (!myNick) return false
  return message.nick.toLowerCase() === myNick.toLowerCase()
}

/**
 * The message the Up key opens for editing, or null if there is none.
 *
 * The most recent one, which is what somebody pressing Up means — the thing
 * they just said and got wrong. Ordered oldest to newest, as a conversation
 * is held everywhere else in this codebase.
 */
export function lastEditable(
  messages: readonly EditableMessage[],
  myNick: string,
  capabilities: readonly string[]
): EditableMessage | null {
  for (let i = messages.length - 1; i >= 0; i--) {
    if (canEdit(messages[i], myNick, capabilities)) return messages[i]
  }
  return null
}
